pkg/storegateway/indexheader/index/postings.go

```diff
@@ -57,93 +57,139 @@ func newV1PostingOffsetTable(factory *streamencoding.DecbufFactory, tableOffset
+
+	startLen := d.Len()
+	remainingCount := d.Be32()
+	currentName := ""
+	valuesForCurrentKey := 0
+	lastEntryOffsetInTable := -1
+
+	for d.Err() == nil && remainingCount > 0 {
+		lastName := currentName
+		offsetInTable := startLen - d.Len()
+		keyCount := d.Uvarint()
 
-	// For the postings offset table we keep every label name but only every nth
-	// label value (plus the first and last one), to save memory.
-	if err := readOffsetTable(factory, tableOffset, func(key []string, off uint64, tableOff int) error {
-		if len(key) != 2 {
-			return errors.Errorf("unexpected key length for posting table %d", len(key))
+		// The Postings offset table takes only 2 keys per entry (name and value of label).
+		if keyCount != 2 {
+			return nil, errors.Errorf("unexpected key length for posting table %d", keyCount)
 		}
 
-		if _, ok := t.postings[key[0]]; !ok {
-			// Not seen before label name.
-			t.postings[key[0]] = &postingValueOffsets{}
-			if lastKey != nil {
-				// Always include last value for each label name, unless it was just added in previous iteration based
-				// on valueCount.
-				if (valueCount-1)%postingOffsetsInMemSampling != 0 {
-					t.postings[lastKey[0]].offsets = append(t.postings[lastKey[0]].offsets, postingOffset{value: lastKey[1], tableOff: lastTableOff})
-				}
-				t.postings[lastKey[0]].lastValOffset = int64(off - crc32.Size)
-				lastKey = nil
+		// Important: this value is only valid as long as we don't perform any further reads from d.
+		// If we need to retain its value, we must copy it before performing another read.
+		unsafeName := d.UnsafeUvarintBytes()
+
+		if len(t.postings) == 0 || currentName != string(unsafeName) {
+			newKey := string(unsafeName)
+
+			if lastEntryOffsetInTable != -1 {
+				// We haven't recorded the last offset for the last value of the previous name.
+				// Go back and read the last value for the previous name.
+				newValueOffset := d.Len()
+				d.ResetAt(lastEntryOffsetInTable + 4) // 4 bytes for entry count
+				d.Uvarint()                           // Skip the key count
+				d.SkipUvarintBytes()                  // Skip the name
+				value := d.UvarintStr()
+				t.postings[currentName].offsets = append(t.postings[currentName].offsets, postingOffset{value: value, tableOff: lastEntryOffsetInTable})
+
+				// Skip ahead to where we were before we called ResetAt() above.
+				d.Skip(d.Len() - newValueOffset)
 			}
-			valueCount = 0
+
+			currentName = newKey
+			t.postings[currentName] = &postingValueOffsets{}
+			valuesForCurrentKey = 0
 		}
 
-		lastKey = key
-		lastTableOff = tableOff
-		valueCount++
+		// Retain every 1-in-postingOffsetsInMemSampling entries, starting with the first one.
+		if valuesForCurrentKey%postingOffsetsInMemSampling == 0 {
+			value := d.UvarintStr()
+			off := d.Uvarint64()
+			t.postings[currentName].offsets = append(t.postings[currentName].offsets, postingOffset{value: value, tableOff: offsetInTable})
```

Comment:
[nit] We're updating `currentName` which is the "previous name" because `currentName` hasn't been updated yet. Can we find a less tricky way to name variables so that reading this code is a bit more straightforward? Maybe just using `lastName` and in the `if` condition entering this branch is enough to address my feedback?

Reply:
At this point, `currentName` has been updated - it would be changed on line 136 above if we've just started reading a new name.

Reply:
I agree. I just wrote my comment on the wrong line 🤦 My comment was about the `currentName` usage at line 130.

Reply:
I've opened https://github.com/grafana/mimir/pull/3797 to address this.